File: conductor/lib/conductor/pipeline.ex
defmodule Conductor.Pipeline do
  @moduledoc """
  Agent pipeline — pure functions for phase progression.

  Mirrors dispatch.rs agent_pipeline/default_agent/next_agent.
  """

  @pipelines %{
    "bug" => ["dev-agent", "staging-agent"],
    "feature" => ["dev-agent", "staging-agent", "prod-agent"],
    "task" => ["dev-agent"],
    "chore" => ["dev-agent"],
    "review" => ["staging-agent"],
    "epic" => ["pm-agent"],
    "design" => ["pm-agent"],
    "research" => ["pm-agent"]
  }

  @doc "The agent pipeline for a given issue type."
  def pipeline(issue_type) do
    Map.get(@pipelines, issue_type, ["dev-agent"])
  end

  @doc "The default (first) agent for a given issue type."
  def default_agent(issue_type) do
    pipeline(issue_type) |> List.first("dev-agent")
  end

  @doc "The next agent after `current`, or nil if pipeline complete."
  def next_agent(issue_type, current) do
    agents = pipeline(issue_type)

    case Enum.find_index(agents, &(&1 == current)) do
      nil -> nil
      idx -> Enum.at(agents, idx + 1)
    end
  end

  @doc "Is this the final phase?"
  def final_phase?(issue_type, current) do
    next_agent(issue_type, current) == nil
  end

  @doc "Total number of phases for an issue type."
  def phase_count(issue_type), do: length(pipeline(issue_type))
end

File: src/dispatch/pipeline.rs
//! Agent pipeline — phase progression and agent directory resolution.

use std::path::PathBuf;

use crate::scanner::expand_path;

/// The agent pipeline for a given issue type.
pub fn agent_pipeline(issue_type: &str) -> &'static [&'static str] {
    match issue_type {
        "bug" => &["dev-agent", "staging-agent"],
        "feature" => &["dev-agent", "staging-agent", "prod-agent"],
        "task" | "chore" => &["dev-agent"],
        "review" => &["staging-agent"],
        "design" | "research" => &["architect-agent"],
        "epic" => &["pm-agent"],
        _ => &["dev-agent"],
    }
}

/// The default (first) agent for a given issue type.
pub fn default_agent(issue_type: &str) -> &'static str {
    agent_pipeline(issue_type)
        .first()
        .copied()
        .unwrap_or("dev-agent")
}

/// The next agent in the pipeline after `current`, or None if done.
pub fn next_agent(issue_type: &str, current: &str) -> Option<&'static str> {
    let pipeline = agent_pipeline(issue_type);
    let idx = pipeline.iter().position(|&a| a == current)?;
    pipeline.get(idx + 1).copied()
}

/// Resolve agents_dir from config by finding the self-managed repo.
pub fn resolve_agents_dir() -> Option<PathBuf> {
    let cfg = crate::config::load_global().ok()?;
    cfg.repo
        .iter()
        .find(|r| r.self_managed)
        .map(|r| expand_path(&r.path).join("agents"))
        .filter(|p| p.exists())
}
